File: src/dagreach/loading.py
"""One door for every input: pick a profile, read, orient.

The order matters and is the same for every command. The profile decides how to
read the file and which way its edges point; an explicit flag always wins over
the profile; and whatever ends up applied is stated in the report rather than
assumed by the reader.
"""

from __future__ import annotations

from dagreach.adapters import PROFILES, Profile, detect_profile, get_profile
from dagreach.model import Graph
from dagreach.readers import read_source
from dagreach.semantics import DEFAULT_SEMANTICS, orient, warn_if_orientation_is_suspect


def load_graph(
    path: str,
    *,
    profile: str | None = None,
    format: str | None = None,
    edge_semantics: str | None = None,
) -> Graph:
    """Read `path` through a profile — named, detected, or generic — and orient it."""
    text, source = read_source(path)

    chosen, detected = _resolve_profile(text, profile)
    graph = (
        chosen.load(text, source)
        if chosen.name != "generic"
        else _generic(text, source, format, path)
    )
    graph.profile = chosen.name

    if format and chosen.name != "generic":
        graph.warn(
            f"--format {format} was ignored: the {chosen.name} profile knows the format it reads"
        )
    if detected:
        graph.warn(
            f"read with the {chosen.name} profile, recognised from the file itself; "
            "pass --profile to choose explicitly"
        )

    semantics = edge_semantics or chosen.edge_semantics or DEFAULT_SEMANTICS
    graph = orient(graph, semantics)
    if edge_semantics and chosen.name != "generic" and edge_semantics != chosen.edge_semantics:
        graph.warn(
            f"--edge-semantics {edge_semantics} overrides the {chosen.name} profile, "
            f"which declares {chosen.edge_semantics}"
        )
    if chosen.name == "generic":
        warn_if_orientation_is_suspect(graph, edge_semantics)
    return graph


def _resolve_profile(text: str, requested: str | None) -> tuple[Profile, bool]:
    if requested:
        return get_profile(requested), False
    found = detect_profile(text)
    if found is not None:
        return found, True
    return PROFILES["generic"], False


def _generic(text: str, source: str, format: str | None, hint: str) -> Graph:
    from dagreach.readers import read_text

    return read_text(text, source=source, format=format, hint=hint)

File: internal/dagreach/loading.go
package dagreach

// One door for every input: pick a profile, read, orient.
//
// The order matters and is the same for every command. The profile decides how
// to read the file and which way its edges point; an explicit flag always wins
// over the profile; and whatever ends up applied is stated in the report rather
// than assumed by the reader.

import (
	"fmt"
	"io"
)

// LoadOptions are the three things a caller can force.
type LoadOptions struct {
	Profile       string
	Format        string
	EdgeSemantics string
	Stdin         io.Reader
}

// LoadGraph reads path through a profile - named, detected, or generic - and orients it.
func LoadGraph(path string, options LoadOptions) (*Graph, error) {
	text, source, err := ReadSource(path, options.Stdin)
	if err != nil {
		return nil, err
	}

	chosen, detected, err := resolveProfile(text, options.Profile)
	if err != nil {
		return nil, err
	}

	var graph *Graph
	if chosen.Name == "generic" {
		graph, err = ReadText(text, source, options.Format, path)
	} else {
		graph, err = chosen.Load(text, source)
	}
	if err != nil {
		return nil, err
	}
	graph.Profile = chosen.Name

	if options.Format != "" && chosen.Name != "generic" {
		graph.Warn(fmt.Sprintf(
			"--format %s was ignored: the %s profile knows the format it reads",
			options.Format, chosen.Name))
	}
	if detected {
		graph.Warn(fmt.Sprintf(
			"read with the %s profile, recognised from the file itself; "+
				"pass --profile to choose explicitly", chosen.Name))
	}

	semantics := options.EdgeSemantics
	if semantics == "" {
		semantics = chosen.EdgeSemantics
	}
	if semantics == "" {
		semantics = DefaultSemantics
	}
	if !knownSemantics(semantics) {
		return nil, &UsageError{Message: fmt.Sprintf("unknown edge semantics '%s'", semantics)}
	}
	graph = Orient(graph, semantics)

	if options.EdgeSemantics != "" && chosen.Name != "generic" &&
		options.EdgeSemantics != chosen.EdgeSemantics {
		graph.Warn(fmt.Sprintf(
			"--edge-semantics %s overrides the %s profile, which declares %s",
			options.EdgeSemantics, chosen.Name, chosen.EdgeSemantics))
	}
	if chosen.Name == "generic" {
		WarnIfOrientationIsSuspect(graph, options.EdgeSemantics)
	}
	// Nodes and no edges is a legitimate answer to `parse` and a trap for
	// everything else: nothing reaches anything, so every reach policy passes
	// without judging a thing. Said once here rather than in each reader,
	// because every reader can produce one.
	if graph.NodeCount() > 0 && graph.EdgeCount() == 0 {
		graph.Warn("this graph has no edges, so nothing reaches anything and every " +
			"reach policy passes without judging anything")
	}
	return graph, nil
}

func resolveProfile(text, requested string) (*Profile, bool, error) {
	if requested != "" {
		profile, ok := GetProfile(requested)
		if !ok {
			return nil, false, &UsageError{Message: fmt.Sprintf("unknown profile '%s'", requested)}
		}
		return profile, false, nil
	}
	if found := DetectProfile(text); found != nil {
		return found, true, nil
	}
	generic, _ := GetProfile("generic")
	return generic, false, nil
}
